---
id: 2045c3
name: alert Role or Live Region Identify Input Error
rule_type: atomic

description: |
  This rule checks that a container with `role=alert` or a live region are used to identify input errors

accessibility_requirements: # Remove whatever is not applicable
  wcag-technique:ARIA19: # Using ARIA role=alert or Live Regions to Identify Errors
    forConformance: false
    failed: not satisfied
    passed: satisfied
    inapplicable: further testing needed

input_aspects:
  - DOM Tree
  - CSS Styling
  - Language

authors:
  - Carlos Duarte
  - Jo√£o Vicente
---

## Applicability

This rule applies to any [document](https://www.w3.org/TR/dom/#concept-document) where the [document element](https://www.w3.org/TR/dom/#document-element) is an HTML `html` element that has at least one [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element).

## Expectation 1

The target element contains an HTML element with either `role=alert` or `aria-live=assertive` attribute that is present in the [document](https://www.w3.org/TR/dom/#concept-document) at the time it [completes loading](https://www.w3.org/TR/html52/dom.html#dom-documentreadystate-complete).

## Expectation 2

The HTML element with either `role=alert` or `aria-live=assertive` attribute displays a message when an [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element) is [completed](#completed-input-field) without meeting the instructions provided for it.

Note: Instructions for an [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element) include information that is required by the Web page, the data format and possible values. The instructions can be presented to the user in several ways, including:

- Text content placed visually in the vicinity of the [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element)
- The accessible description of the [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element)
- A tooltip displayed when the [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element) is [focused](#focusable)
- Text content accessible through an help button or similar

## Expectation 3

The content of the message is [visible](#visible) and [included in the accessibility tree](included-in-the-accessibility-tree), and identifies the [input error](https://www.w3.org/TR/WCAG21/#dfn-input-error).

## Assumptions

_There are currently no assumptions_

## Accessibility Support

_There are no major accessibility support issues known for this rule._

## Background

- [ARIA19: Using ARIA role=alert or Live Regions to Identify Errors](https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA19)

## Test Cases

### Passed

#### Passed Example 1

Description...

```html
<!-- code -->
```

#### Passed Example 2

...

### Failed

#### Failed Example 1

Description...

```html
<!-- code -->
```

#### Failed Example 2

...

### Inapplicable

#### Inapplicable Example 1

Description...

```html
<!-- code -->
```

#### Inapplicable Example 2

...