---
id: 334972
name: Required form fields not completed
rule_type: atomic

description: |
  This rule checks that text error messages identify required form fields that were not completed

accessibility_requirements: # Remove whatever is not applicable
  wcag-technique:G83: # Providing text descriptions to identify required fields that were not completed
    forConformance: false
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed

input_aspects:
  - DOM Tree # (e.g. HTTP Messages, DOM Tree, CSS Styling, Accessibility Tree, Language, etc.,)
  - CSS Styling
  - Language

authors:
  - Carlos Duarte
  - Jo√£o Vicente
---

## Applicability

The rule applies to each [form element](https://www.w3.org/TR/html52/sec-forms.html#the-form-element) that includes at least one [required](#required-input-field) [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element).

## Expectation 1

After triggering the submission of the target element, each [required](#required-input-field) [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element) that was not [completed](#completed-input-field) is identified by a text message. 

Note: A text message may identify an [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element) in several ways, including:

- By referring to its label or accessible name
- By being visually placed in the vicinity of the element

## Expectation 2

The content of text message, , [visible](#visible) and [included in the accessibility tree](included-in-the-accessibility-tree), indicates that the [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element) is [required](#required-input-field).

## Expectation 3

All [completed fields](#completed-input-fields) of the target element retain their values.

## Assumptions

_There are currently no assumptions_

## Accessibility Support

_There are no major accessibility support issues known for this rule._

## Background

- [G83: Providing text descriptions to identify required fields that were not completed](https://www.w3.org/WAI/WCAG21/Techniques/general/G83)

## Test Cases

### Passed

#### Passed Example 1

Description...

```html
<!-- code -->
```

#### Passed Example 2

...

### Failed

#### Failed Example 1

Description...

```html
<!-- code -->
```

#### Failed Example 2

...

### Inapplicable

#### Inapplicable Example 1

Description...

```html
<!-- code -->
```

#### Inapplicable Example 2

...