---
id: 36b590
name: Invalid form field value
rule_type: atomic

description: |
  This rule checks that text descriptions are provided when the user completes a form field with information that is not an allowed value or using a not allowed format

accessibility_requirements: # Remove whatever is not applicable
  wcag-technique:G84: # Providing a text description when the user provides information that is not in the list of allowed values
    forConformance: false
    failed: not satisfied
    passed: satisfied
    inapplicable: further testing needed

  wcag-technique:G85: # Providing a text description when user input falls outside the required format or values
    forConformance: false
    failed: not satisfied
    passed: satisfied
    inapplicable: further testing needed

input_aspects:
  - DOM Tree
  - CSS Styling
  - Language

authors:
  - Carlos Duarte
  - Jo√£o Vicente
---

## Applicability

The rule applies to each [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element), that may or may not belong to a [form element](https://www.w3.org/TR/html52/sec-forms.html#the-form-element).
## Expectation 1

After [user completion](#completed-input-field) of the target element or triggering the submission of the form, if the target element belongs to one, if the target element does not meet the instructions provided for it, a message is presented to the user.

Note: Instructions for an [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element) include information that is required by the Web page, the data format and possible values. The instructions can be presented to the user in several ways, including:

- Text content placed visually in the vicinity of the [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element)
- The accessible description of the [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element)
- A tooltip displayed when the [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element) is [focused](#focusable)
- Text content accessible through an help button or similar

## Expectation 2

The content of the message is [visible](#visible) and [included in the accessibility tree](included-in-the-accessibility-tree), and identifies the [input error](https://www.w3.org/TR/WCAG21/#dfn-input-error).

## Assumptions

_There are currently no assumptions_

## Accessibility Support

_There are no major accessibility support issues known for this rule._

## Background

- [G84: Providing a text description when the user provides information that is not in the list of allowed values](https://www.w3.org/WAI/WCAG21/Techniques/general/G84)
- [G85: Providing a text description when user input falls outside the required format or values](https://www.w3.org/WAI/WCAG21/Techniques/general/G85)

## Test Cases

### Passed

#### Passed Example 1

Description...

```html
<!-- code -->
```

#### Passed Example 2

...

### Failed

#### Failed Example 1

Description...

```html
<!-- code -->
```

#### Failed Example 2

...

### Inapplicable

#### Inapplicable Example 1

Description...

```html
<!-- code -->
```

#### Inapplicable Example 2

...