---
id: 54621b
name: aria-invalid Identifies Input Error
rule_type: atomic

description: |
  This rule checks that the `aria-invalid` attributed is used to identify input errors

accessibility_requirements: # Remove whatever is not applicable
  wcag-technique: ARIA21: # Using Aria-Invalid to Indicate An Error Field
    forConformance: false
    failed: not satisfied
    passed: satisfied
    inapplicable: further testing needed

input_aspects:
  - DOM Tree
  - CSS Styling
  - Language

authors:
  - Carlos Duarte
  - Jo√£o Vicente
---

## Applicability

The rule applies to each [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element), that may or may not belong to a [form element](https://www.w3.org/TR/html52/sec-forms.html#the-form-element).

## Expectation 1

After [user completion](#completed-input-field) of the target element or triggering the submission of the form, if the target element belongs to one, each target element that meets the instructions provided for it does not have the `aria-invalid` attribute set to `true`.

Note: Instructions for an [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element) include information that is required by the Web page, the data format and possible values. The instructions can be presented to the user in several ways, including:

- Text content placed visually in the vicinity of the [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element)
- The accessible description of the [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element)
- A tooltip displayed when the [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element) is [focused](#focusable)
- Text content accessible through an help button or similar

## Expectation 2

After [user completion](#completed-input-field) of the target element or triggering the submission of the form, if the target element belongs to one, each target element that does not meet the instructions provided for it has the `aria-invalid` attribute set to `true`.

## Expectation 3

Each target element that does not meet the instructions provided for it has a [programmatically determined](https://www.w3.org/TR/WCAG21/#dfn-programmatically-determinable) label or text instruction identifying the [input error](https://www.w3.org/TR/WCAG21/#dfn-input-error).

## Assumptions

_There are currently no assumptions_

## Accessibility Support

_There are no major accessibility support issues known for this rule._

## Background

- [ARIA21: Using Aria-Invalid to Indicate An Error Field](https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA21)

## Test Cases

### Passed

#### Passed Example 1

Description...

```html
<!-- code -->
```

#### Passed Example 2

...

### Failed

#### Failed Example 1

Description...

```html
<!-- code -->
```

#### Failed Example 2

...

### Inapplicable

#### Inapplicable Example 1

Description...

```html
<!-- code -->
```

#### Inapplicable Example 2

...