---
id: 6f484a
name: aria-alertdialog Identifies Input Error
rule_type: atomic

description: |
  This rule checks that `aria-alertdialog` is used to identify input errors

accessibility_requirements: # Remove whatever is not applicable
  wcag-technique:ARIA18: # Using aria-alertdialog to Identify Errors
    forConformance: false
    failed: not satisfied
    passed: satisfied
    inapplicable: further testing needed

input_aspects:
  - DOM Treee
  - CSS Styling
  - Language

authors:
  - Carlos Duarte
  - Jo√£o Vicente
---

## Applicability

The rule applies to each [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element).

## Expectation 

After [user completion](#completed-input-field) of the target element or triggering the submission of the form, if the target element belongs to one, each target element that does not meet the instructions provided for it causes an element with `role="alertdialog"` to appear that
- contains at least one [focusable](#focusable) element; and
- the focus moves to the [focusable](#focusable) element; and
- it is not possible to move the focus away from elements contained in the element with `role="alertdialog"` until this is dismissed; and
- after the element with `role="alertdialog"` is dismissed the focus returns to the position where it was before the element with `role="alertdialog"` was displayed; and
- the element with `role="alertdialog"` has an [accessible name](accessible-name); and
- the content of the element with `role="alertdialog"` identifies the [input error](https://www.w3.org/TR/WCAG21/#dfn-input-error).

Note: Instructions for an [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element) include information that is required by the Web page, the data format and possible values. The instructions can be presented to the user in several ways, including:

- Text content placed visually in the vicinity of the [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element)
- The accessible description of the [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element)
- A tooltip displayed when the [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element) is [focused](#focusable)
- Text content accessible through an help button or similar

## Assumptions

_There are currently no assumptions_

## Accessibility Support

_There are no major accessibility support issues known for this rule._

## Background

- [ARIA18: Using aria-alertdialog to Identify Errors](https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA18)

## Test Cases

### Passed

#### Passed Example 1

Description...

```html
<!-- code -->
```

#### Passed Example 2

...

### Failed

#### Failed Example 1

Description...

```html
<!-- code -->
```

#### Failed Example 2

...

### Inapplicable

#### Inapplicable Example 1

Description...

```html
<!-- code -->
```

#### Inapplicable Example 2

...